### PR TITLE
mgr,mon: subscribe to config-key changes from mgr; maintain mgr-side ConfigMap

### PR DESCRIPTION
**Pull request by liewegas**
- convert mon ConfigKeyService to KVMonitor (PaxosService)
- allow subscriptions to config-key prefixes
- make mgr subscribe to mgr/, device/.  This results in cleaner/faster startup and allows mgr to see config-key changes made via the CLI
- make mgr subscribe to config/ and maintain a ConfigMap
- implement mgr_module get_foreign_ceph_option(entity, name) to get any ceph config option for any daemon (excluding mgr module options)
- update cephadm to use get_foreign_ceph_option() instead of 'config get'

For the most part there are no upgrade/downgrade compat issues here, except that a new mgr may be running against a mon cluster that doesn't (all) support kv subscriptions, in which case it still needs to do the old slow thing of fetching all of the config-key values individually.

### CODE REVIEW

File: src/mon/KVMonitor.cc [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+}
+
+void KVMonitor::update_from_paxos(bool *need_bootstrap)
+{
+  if (version == get_last_committed()) {
+    return;
+  }
+  version = get_last_committed();
+  dout(10) << __func__ << " " << version << dendl;
+  //check_all_subs();
```
Comment (jdurgin) [inline src/mon/KVMonitor.cc]: commented out code

File: src/mon/KVMonitor.cc [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+  }
+  string format;
+  cmd_getval(cmdmap, "format", format, string("plain"));
+  boost::scoped_ptr<Formatter> f(Formatter::create(format));
+
+  string prefix;
+  cmd_getval(cmdmap, "prefix", prefix);
+  string key;
+  cmd_getval(cmdmap, "key", key);
+
```
Comment (jdurgin) [inline src/mon/KVMonitor.cc:154]: are permissions already checked elsewhere? I'd expect those checks here

Comment (jdurgin) [reply src/mon/KVMonitor.cc:154]: also in prepare_command()

File: src/mgr/Mgr.cc [hunk 3]
```diff
@@ -251,6 +235,10 @@ void Mgr::init()
   monc->sub_want("mgrdigest", 0, 0);
   monc->sub_want("fsmap", 0, 0);
   monc->sub_want("servicemap", 0, 0);
+  monc->sub_want("kv:config/", 0, 0);
+  monc->sub_want("kv:mgr/", 0, 0);
+  monc->sub_want("kv:device/", 0, 0);
+  monc->sub_want("kv:config/", 0, 0);
```
Comment (jdurgin) [inline src/mgr/Mgr.cc:241]: duplicate ```kv:config/```

File: src/mon/ConfigMonitor.cc [hunk 4]
```diff
@@ -811,7 +811,7 @@ void ConfigMonitor::load_config()
     if (!opt) {
       dout(10) << __func__ << " unrecognized option '" << name << "'" << dendl;
       opt = new Option(name, Option::TYPE_STR, Option::LEVEL_UNKNOWN);
-      // FIXME: this will be leaked!
+      config_map.stray_options.push_back(opt);
```
Comment (jdurgin) [inline src/mon/ConfigMonitor.cc:805]: it'd be great to get rid of raw pointers here

File: src/mon/KVMonitor.cc (hunk 2 above)
Comment (liewegas) [reply src/mon/KVMonitor.cc:154]: they're checked in the caller at https://github.com/ceph/ceph/blob/master/src/mon/Monitor.cc#L3422

File: src/mgr/Mgr.cc (hunk 3 above)
Comment (liewegas) [reply src/mgr/Mgr.cc:241]: fixed!

File: src/pybind/mgr/cephadm/upgrade.py [hunk 5]
```diff
@@ -511,6 +511,12 @@ def _do_upgrade(self):
                     })
                 return
 
+            # complete mon upgrade?
+            if daemon_type == 'mon':
+                if not self.mgr.get("have_local_config_map"):
+                    logger.info('Upgrade: Restarting mgr now that mons are runnig pacific')
```
Comment (neha-ojha) [inline src/pybind/mgr/cephadm/upgrade.py:517]: I seem to recall we had some issues in the past due to which the mgr needed to be upgraded first, wondering if that doesn't hold any longer.

**Review by jecluis — APPROVED** [review]
It looks good to me.

**Pull request merged**